File: java/tutorial/main/ch/softappeal/yass/tutorial/py/SocketServer.java
package ch.softappeal.yass.tutorial.py;

import ch.softappeal.yass.Exceptions;
import ch.softappeal.yass.Interceptor;
import ch.softappeal.yass.NamedThreadFactory;
import ch.softappeal.yass.Nullable;
import ch.softappeal.yass.remote.Server;
import ch.softappeal.yass.remote.Service;
import ch.softappeal.yass.transport.MessageSerializer;
import ch.softappeal.yass.transport.socket.SimpleSocketTransport;
import ch.softappeal.yass.transport.socket.SocketBinder;
import ch.softappeal.yass.tutorial.contract.EchoService;
import ch.softappeal.yass.tutorial.contract.Instrument;
import ch.softappeal.yass.tutorial.contract.SystemException;
import ch.softappeal.yass.tutorial.contract.instrument.InstrumentService;
import ch.softappeal.yass.tutorial.shared.Logger;
import ch.softappeal.yass.tutorial.shared.SslConfig;
import ch.softappeal.yass.tutorial.shared.socket.SocketSetup;

import javax.net.ssl.SSLSocket;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import static ch.softappeal.yass.tutorial.contract.Config.PY_ACCEPTOR;

public final class SocketServer {

    private static final class EchoServiceImpl implements EchoService {
        @Override public @Nullable Object echo(final @Nullable Object value) {
            if ("exception".equals(value)) {
                throw new SystemException("exception");
            }
            return value;
        }
    }

    private static final class InstrumentServiceImpl implements InstrumentService {
        @Override public List<Instrument> getInstruments() {
            return List.of();
        }
        @Override public void showOneWay(final boolean testBoolean, final int testInt) {
            // empty
        }
    }

    private static final Interceptor LOGGER = new Logger(null, Logger.Side.SERVER);

    private static final Interceptor PEER = (method, arguments, invocation) -> {
        System.out.println(((SSLSocket)SimpleSocketTransport.socket()).getSession().getPeerPrincipal().getName());
        return invocation.proceed();
    };

    public static void main(final String... args) {
        final Executor executor = Executors.newCachedThreadPool(new NamedThreadFactory("executor", Exceptions.STD_ERR));
        new SimpleSocketTransport(
            executor,
            new MessageSerializer(SocketClient.SERIALIZER),
            new Server(
                new Service(PY_ACCEPTOR.echoService, new EchoServiceImpl()),
                new Service(PY_ACCEPTOR.instrumentService, new InstrumentServiceImpl(), PEER, LOGGER)
            )
        ).start(
            executor,
            SocketBinder.create(SslConfig.SERVER.serverSocketFactory, SocketSetup.ADDRESS)
        );
        System.out.println("started");
    }

}

File: kotlin/tutorial/main/ch/softappeal/yass/tutorial/py/SocketServer.kt
package ch.softappeal.yass.tutorial.py

import ch.softappeal.yass.Interceptor
import ch.softappeal.yass.StdErr
import ch.softappeal.yass.namedThreadFactory
import ch.softappeal.yass.remote.Server
import ch.softappeal.yass.remote.Service
import ch.softappeal.yass.transport.ServerSetup
import ch.softappeal.yass.transport.messageSerializer
import ch.softappeal.yass.transport.socket.serverSocketFactory
import ch.softappeal.yass.transport.socket.socket
import ch.softappeal.yass.transport.socket.socketBinder
import ch.softappeal.yass.transport.socket.socketServer
import ch.softappeal.yass.tutorial.contract.EchoService
import ch.softappeal.yass.tutorial.contract.Instrument
import ch.softappeal.yass.tutorial.contract.PyAcceptor
import ch.softappeal.yass.tutorial.contract.SystemException
import ch.softappeal.yass.tutorial.contract.instrument.InstrumentService
import ch.softappeal.yass.tutorial.shared.Side
import ch.softappeal.yass.tutorial.shared.SslServer
import ch.softappeal.yass.tutorial.shared.logger
import ch.softappeal.yass.tutorial.shared.socket.Address
import java.util.concurrent.Executors
import javax.net.ssl.SSLSocket

private val EchoServiceImpl = object : EchoService {
    override fun echo(value: Any?): Any? {
        if ("exception" == value) throw  SystemException("exception")
        return value
    }
}

private val InstrumentServiceImpl = object : InstrumentService {
    override val instruments: List<Instrument> get() = emptyList()
    override fun showOneWay(testBoolean: Boolean, testInt: Int) {}
}

private val Peer: Interceptor = { _, _, invocation ->
    println((socket as SSLSocket).session.peerPrincipal.name)
    invocation()
}

fun main(args: Array<String>) {
    val executor = Executors.newCachedThreadPool(namedThreadFactory("executor", StdErr))
    socketServer(
        ServerSetup(
            Server(
                Service(PyAcceptor.echoService, EchoServiceImpl),
                Service(PyAcceptor.instrumentService, InstrumentServiceImpl, Peer, logger(null, Side.Server))
            ),
            messageSerializer(Serializer)
        ),
        executor
    ).start(
        executor,
        socketBinder(Address, SslServer.serverSocketFactory)
    )
    println("started")
}
